packages/go_router/lib/src/match.dart

```diff
@@ -102,6 +312,116 @@ class RouteMatch {
 
   @override
   int get hashCode => Object.hash(route, matchedLocation, pageKey);
+
+  @override
+  GoRouterState buildState(
+      RouteConfiguration configuration, RouteMatchList matches) {
+    return GoRouterState(
+      configuration,
+      uri: matches.uri,
+      matchedLocation: matchedLocation,
+      fullPath: matches.fullPath,
+      pathParameters: matches.pathParameters,
+      pageKey: pageKey,
+      name: route.name,
+      path: route.path,
+      extra: matches.extra,
+    );
+  }
+}
+
+/// An matched result by matching a [ShellRoute] against a location.
+///
+/// This is typically created by calling [RouteMatchBase.match].
+@immutable
+class ShellRouteMatch extends RouteMatchBase {
+  /// Create a match.
+  ShellRouteMatch({
+    required this.route,
+    required this.matches,
+    required this.matchedLocation,
+    required this.pageKey,
+    required this.navigatorKey,
+  }) : assert(matches.isNotEmpty);
+
+  @override
+  final ShellRouteBase route;
+
+  RouteMatch get _leafMatch {
```

Comment:
I found the name `leafMatch` a bit confusing cause a tree can have many leaves. Maybe `lastLeaf` is more clear?